File: bubus-ts/tests/bridge_listener_worker.ts
import { readFileSync, writeFileSync } from 'node:fs'

import {
  HTTPEventBridge,
  JSONLEventBridge,
  NATSEventBridge,
  PostgresEventBridge,
  RedisEventBridge,
  SQLiteEventBridge,
  SocketEventBridge,
} from '../src/index.js'

type WorkerConfig = {
  kind: string
  ready_path: string
  output_path: string
  endpoint?: string
  path?: string
  table?: string
  url?: string
  server?: string
  subject?: string
}

const makeListenerBridge = (config: WorkerConfig): any => {
  if (config.kind === 'http') return new HTTPEventBridge({ listen_on: config.endpoint })
  if (config.kind === 'socket') return new SocketEventBridge(config.path)
  if (config.kind === 'jsonl') return new JSONLEventBridge(config.path ?? '', 0.05)
  if (config.kind === 'sqlite') return new SQLiteEventBridge(config.path ?? '', config.table ?? 'bubus_events', 0.05)
  if (config.kind === 'redis') return new RedisEventBridge(config.url ?? '')
  if (config.kind === 'nats') return new NATSEventBridge(config.server ?? '', config.subject ?? '')
  if (config.kind === 'postgres') return new PostgresEventBridge(config.url ?? '')
  throw new Error(`Unsupported bridge kind: ${config.kind}`)
}

const main = async (): Promise<void> => {
  const config_path = process.argv[2]
  const config = JSON.parse(readFileSync(config_path, 'utf8')) as WorkerConfig
  const bridge = makeListenerBridge(config)

  let resolve_done: (() => void) | null = null
  const done = new Promise<void>((resolve) => {
    resolve_done = resolve
  })

  await bridge.start()
  bridge.on('*', (event: { toJSON: () => unknown }) => {
    writeFileSync(config.output_path, JSON.stringify(event.toJSON()), 'utf8')
    resolve_done?.()
  })
  writeFileSync(config.ready_path, 'ready', 'utf8')
  await Promise.race([done, new Promise((_, reject) => setTimeout(() => reject(new Error('worker timeout')), 30000))])
  await bridge.close()
}

await main()

File: tests/bridge_listener_worker.py
from __future__ import annotations

import asyncio
import json
import sys
from typing import Any

from anyio import Path as AnyPath

from bubus import HTTPEventBridge, SocketEventBridge
from bubus.bridge_jsonl import JSONLEventBridge
from bubus.bridge_nats import NATSEventBridge
from bubus.bridge_postgres import PostgresEventBridge
from bubus.bridge_redis import RedisEventBridge
from bubus.bridge_sqlite import SQLiteEventBridge


def _make_listener_bridge(config: dict[str, Any]) -> Any:
    kind = str(config['kind'])
    if kind == 'http':
        return HTTPEventBridge(listen_on=str(config['endpoint']))
    if kind == 'socket':
        return SocketEventBridge(path=str(config['path']))
    if kind == 'jsonl':
        return JSONLEventBridge(str(config['path']), poll_interval=0.05)
    if kind == 'sqlite':
        return SQLiteEventBridge(str(config['path']), str(config['table']), poll_interval=0.05)
    if kind == 'redis':
        return RedisEventBridge(str(config['url']))
    if kind == 'nats':
        return NATSEventBridge(str(config['server']), str(config['subject']))
    if kind == 'postgres':
        return PostgresEventBridge(str(config['url']))
    raise ValueError(f'Unsupported bridge kind: {kind}')


async def _main(config_path: str) -> None:
    config = json.loads(await AnyPath(config_path).read_text(encoding='utf-8'))
    ready_path = AnyPath(str(config['ready_path']))
    output_path = AnyPath(str(config['output_path']))
    done = asyncio.Event()

    bridge = _make_listener_bridge(config)

    async def _on_event(event: Any) -> None:
        await output_path.write_text(json.dumps(event.model_dump(mode='json')), encoding='utf-8')
        done.set()

    bridge.on('*', _on_event)
    await bridge.start()
    await ready_path.write_text('ready', encoding='utf-8')
    try:
        await asyncio.wait_for(done.wait(), timeout=30.0)
    finally:
        await bridge.close()


if __name__ == '__main__':
    asyncio.run(_main(sys.argv[1]))
